# Penguin Data

## import packages
import numpy as np

impoer pandas as pd

In [2]:
import numpy as np
import pandas as pd

## 1. read DataSet
use pandas

pd.read_csv('addres/name.csv')

In [4]:
data=pd.read_csv('penguins_size.csv')
data.head()

species     island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen              39.1             18.7              181.0   
1  Adelie  Torgersen              39.5             17.4              186.0   
2  Adelie  Torgersen              40.3             18.0              195.0   
3  Adelie  Torgersen               NaN              NaN                NaN   
4  Adelie  Torgersen              36.7             19.3              193.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE

# 2. find out shape of data

In [6]:
data_shaope = data.shape
data_size = data.size
data_ndim = data.ndim

print(f'data_shaope = {data_shaope} \ndata_size = {data_size} \ndata_ndim = {data_ndim}')

data_shaope = (344, 7) 
data_size = 2408 
data_ndim = 2


# 3. dataset info

## Information Of data
in the first print Information of data

there are 344 row and 7 column in the data

there are null value

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   culmen_length_mm   342 non-null    float64
 3   culmen_depth_mm    342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                334 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [18]:
data.head(7)

species     island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen              39.1             18.7              181.0   
1  Adelie  Torgersen              39.5             17.4              186.0   
2  Adelie  Torgersen              40.3             18.0              195.0   
3  Adelie  Torgersen               NaN              NaN                NaN   
4  Adelie  Torgersen              36.7             19.3              193.0   
5  Adelie  Torgersen              39.3             20.6              190.0   
6  Adelie  Torgersen              38.9             17.8              181.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       3625.0  FEMALE

In [19]:
data.tail()

species  island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
339  Gentoo  Biscoe               NaN              NaN                NaN   
340  Gentoo  Biscoe              46.8             14.3              215.0   
341  Gentoo  Biscoe              50.4             15.7              222.0   
342  Gentoo  Biscoe              45.2             14.8              212.0   
343  Gentoo  Biscoe              49.9             16.1              213.0   

     body_mass_g     sex  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE

In [20]:
data["culmen_depth_mm"].describe

<bound method NDFrame.describe of 0      18.7
1      17.4
2      18.0
3       NaN
4      19.3
       ... 
339     NaN
340    14.3
341    15.7
342    14.8
343    16.1
Name: culmen_depth_mm, Length: 344, dtype: float64>

# 4. unique Species

there are null (nan) and Wrong data ( . ) at sex of penguins

In [11]:
# species of penguins
print(f'species of penguins is ----{data['species'].unique()}')
print(f'island of penguins is -----{data['island'].unique()}')
print(f'sex of penguins is --------{data['sex'].unique()}')

species of penguins is ----['Adelie' 'Chinstrap' 'Gentoo']
island of penguins is -----['Torgersen' 'Biscoe' 'Dream']
sex of penguins is --------['MALE' 'FEMALE' nan '.']


# 5.	find NaN values per columns

In [13]:
data[data.isnull().any(axis='columns')]
nan_values = data.isnull().sum()

# 6.	find culmen_depth_mm in each island

In [23]:
data_back_fill = data.bfill()
print(data_back_fill.iloc[3])
print(data.iloc[4])
print(data.iloc[3])

species                 Adelie
island               Torgersen
culmen_length_mm          36.7
culmen_depth_mm           19.3
flipper_length_mm        193.0
body_mass_g             3450.0
sex                     FEMALE
Name: 3, dtype: object
species                 Adelie
island               Torgersen
culmen_length_mm          36.7
culmen_depth_mm           19.3
flipper_length_mm        193.0
body_mass_g             3450.0
sex                     FEMALE
Name: 4, dtype: object
species                 Adelie
island               Torgersen
culmen_length_mm           NaN
culmen_depth_mm            NaN
flipper_length_mm          NaN
body_mass_g                NaN
sex                        NaN
Name: 3, dtype: object


In [19]:
min_culmen_length_mm = data.groupby('island')['culmen_length_mm'].min()
max_culmen_length_mm = data.groupby('island')['culmen_length_mm'].max()
median_culmen_length_mm = data.groupby('island')['culmen_length_mm'].median()
mean_culmen_length_mm = data.groupby('island')['culmen_length_mm'].mean()
print(f'-----min_culmen_length_mm-----\n {min_culmen_length_mm}')
print(f'-----max_culmen_length_mm-----\n {max_culmen_length_mm}')
print(f'-----median_culmen_length_mm-----\n {median_culmen_length_mm}')
print(f'-----mean_culmen_length_mm-----\n {mean_culmen_length_mm}')


-----min_culmen_length_mm-----
 island
Biscoe       34.5
Dream        32.1
Torgersen    33.5
Name: culmen_length_mm, dtype: float64
-----max_culmen_length_mm-----
 island
Biscoe       59.6
Dream        58.0
Torgersen    46.0
Name: culmen_length_mm, dtype: float64
-----median_culmen_length_mm-----
 island
Biscoe       45.80
Dream        44.65
Torgersen    38.90
Name: culmen_length_mm, dtype: float64
-----mean_culmen_length_mm-----
 island
Biscoe       45.257485
Dream        44.167742
Torgersen    38.950980
Name: culmen_length_mm, dtype: float64


In [21]:
culmen_depth_island = data.groupby('island')['culmen_depth_mm'].mean()
culmen_depth_island

island
Biscoe       15.874850
Dream        18.344355
Torgersen    18.429412
Name: culmen_depth_mm, dtype: float64

In [23]:
pivot_table_culmen_depth_mm = pd.pivot_table(data, values='culmen_depth_mm', index='island',
aggfunc= {'culmen_depth_mm': ['min', 'max', 'mean', 'median']} )
pivot_table_culmen_depth_mm

max       mean  median   min
island                                  
Biscoe     21.1  15.874850    15.5  13.1
Dream      21.2  18.344355    18.4  15.5
Torgersen  21.5  18.429412    18.4  15.9

# 7.	body mass in each sex

In [17]:
data_back_fill = data.bfill()
print(f'befor clean data {data_back_fill['sex'].unique()}')
# Change value  sex =  "."
data_back_fill['sex'] = data_back_fill['sex'].replace(".", "MALE")
print(f'after clean data {data_back_fill['sex'].unique()}')

#----------------------------------------
body_mass_sex = data_back_fill.groupby('sex')['body_mass_g'].mean()
print(f'body mass in each sex {body_mass_sex}')

befor clean data ['MALE' 'FEMALE' '.']
after clean data ['MALE' 'FEMALE']
body mass in each sex sex
FEMALE    3856.250000
MALE      4546.656977
Name: body_mass_g, dtype: float64


In [24]:
pivot_table_body_mass_in_each_sex = pd.pivot_table(data, values='body_mass_g', index='sex',
aggfunc= {'body_mass_g': ['min', 'max', 'mean', 'median']} )
pivot_table_body_mass_in_each_sex

max         mean  median     min
sex                                        
.       4875.0  4875.000000  4875.0  4875.0
FEMALE  5200.0  3862.272727  3650.0  2700.0
MALE    6300.0  4545.684524  4300.0  3250.0

# 8.	species in each island

In [25]:
data_back_fill.groupby(["species","island"]).agg("island").count().unstack()


island     Biscoe  Dream  Torgersen
species                            
Adelie       44.0   56.0       52.0
Chinstrap     NaN   68.0        NaN
Gentoo      124.0    NaN        NaN

In [28]:
female_share_species = data[data['sex'] == 'FEMALE'].groupby('species').size() / data.groupby('species').size()
female_share_species

species
Adelie       0.480263
Chinstrap    0.500000
Gentoo       0.467742
dtype: float64

# 9.	share of female penguins in each species

In [29]:
shape_of_females = pd.pivot_table(
    data_back_fill,
    values="island",
    columns="sex",
    index="species",
    aggfunc="count",
    margins=True,).assign(share_female=lambda df: df["FEMALE"]/df["All"])
shape_of_females

sex        FEMALE  MALE  All  share_female
species                                   
Adelie         79    73  152      0.519737
Chinstrap      34    34   68      0.500000
Gentoo         59    65  124      0.475806
All           172   172  344      0.500000

# 10.	flipper_length_mm comparisons between species ( min, max, mean )

In [30]:
comparisons = data_back_fill.groupby('species')['flipper_length_mm'].aggregate(["min", 'max', 'mean'])
comparisons

min    max        mean
species                            
Adelie     172.0  210.0  189.973684
Chinstrap  178.0  212.0  195.823529
Gentoo     203.0  231.0  217.169355